ice-adapter/src/main/java/com/faforever/iceadapter/util/latencyestimation/FafLatencyEstimator.java

```diff
@@ -0,0 +1,92 @@
+package com.faforever.iceadapter.util.latencyestimation;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+
+@Slf4j
+public class FafLatencyEstimator {
+
+    private static final int FAF_LATENCY_ESTIMATION_RETRIES = 3;
+    private static final int FAF_LATENCY_ESTIMATION_RETRY_PERIOD = 500;
+    public static final String FAF_REQUEST = "{\"command\":\"ping\"}\n";
+
+
+    // estimate latency multiple times and take lowest to compensate for server processing time
+    public static CompletableFuture<Double> getLatency(String hostname) {
+        CompletableFuture<Double> future = new CompletableFuture<>();
+
+        log.info("Estimating latency to {} using fafp", hostname);
+
+        new Thread(() -> {
+            List<CompletableFuture<Double>> estimations = new ArrayList<>();
+            for(int i = 0;i < FAF_LATENCY_ESTIMATION_RETRIES;i++) {
+                estimations.add(pingFafServer(hostname));
+                try { Thread.sleep(FAF_LATENCY_ESTIMATION_RETRY_PERIOD); } catch(InterruptedException ignored) {}
+            }
+            CompletableFuture.allOf(estimations.toArray(new CompletableFuture[0])).thenRun(() -> {
+                double lowest = estimations.stream().mapToDouble(doubleCompletableFuture -> {
+                    try {
+                        return doubleCompletableFuture.get();
```

Comment:
why are you using get instead of join with a .exceptionally? 

Also if you use get should probably handle the CancellationException as well

Reply:
I am handeling the cancellation exception, just not separately? Or what did you mean?
About the exceptionally, I'm not too much a fan of that as I would be adding a listener even though I'm already in a block triggered when all futures have completed, so I can just get the result. I don't see the issue with throwing and catching the exception here.